File: Programmers/programmers_신고결과받기.cpp
#include <string>
#include <vector>
#include <sstream>
#include <iostream>
#include <set>
#include <unordered_map>
using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    vector<int> answer(id_list.size(),0);
    unordered_map<string,int> idxs;
    for(int i=0;i<id_list.size();i++){
        idxs[id_list[i]]=i;
    }
    unordered_map<string,set<string>> um;
    stringstream ss;
    for(int i=0;i<report.size();i++){
        ss.str(report[i]);
        string rper, rped;
        ss >> rper >> rped;
        um[rped].insert(rper);
        ss.clear();
    }
    for(auto it=um.begin(); it!= um.end(); it++){
        if(it->second.size()<k) continue;
        for(auto ed_it:it->second){
            int idx = idxs[ed_it];
            answer[idx]++;
        }
    }
    return answer;
}

File: Programmers/programmers_신고결과받기.py
from collections import defaultdict
def solution(id_list, report, k):
    answer = [0] * len(id_list) 
    report = set(report) #신고내용 중복 제거
    
    #A가 신고한 목록 : reports
    #A를 신고한 목록 : reporteds
    reports = defaultdict(list)
    reporteds = defaultdict(list)
    for str in report :
        reporter, reported = str.split()
        if reported not in reports[reporter]:
            reports[reporter].append(reported)
            if reporter not in reporteds[reported]:
                reporteds[reported].append(reporter)
    for idx in range(0,len(id_list)) :
        if len(reporteds[id_list[idx]])>=k:
            for str in reporteds[id_list[idx]]:
                answer[id_list.index(str)]+=1;
     return answer
